src/core/water.cpp

```diff
@@ -226,76 +236,76 @@ void Water::ShakeWaters3d(int number_of_threads) {
 
     RandomNumberGenerator my_rand(local_threads);
 
-    if ( (float)number_of_waters < (powf(2, 31) - 1) / 3.0f ) {
-#pragma omp parallel for num_threads(local_threads) private(dr, dx, dy, dz, my_rand)
-        for ( long iWater = 0; iWater < number_of_waters; iWater++ ) {
+    static constexpr std::size_t block_size = 1024;
+    std::size_t                  num_blocks = (number_of_waters + block_size - 1) / block_size;
 
-            water_coords[iWater].x += my_rand.GetNormalRandomSTD(0.0f, random_sigma);
-            water_coords[iWater].y += my_rand.GetNormalRandomSTD(0.0f, random_sigma);
-            water_coords[iWater].z += my_rand.GetNormalRandomSTD(0.0f, random_sigma);
+    VSLStreamStatePtr stream;
+    vslNewStream(&stream, VSL_BRNG_PHILOX4X32X10, seed_++);
 
-            // TODO 2x check that the periodic shifts are doing what they should be.
-            // Check boundaries
-            if ( water_coords[iWater].x < size_neighborhood + 1 ) {
-                water_coords[iWater].x = water_coords[iWater].x - 1 * size_neighborhood + vol_nX;
-            }
-            else if ( water_coords[iWater].x > vol_nX - size_neighborhood ) {
-                water_coords[iWater].x = water_coords[iWater].x - vol_nX + 1 * size_neighborhood;
-            }
+#pragma omp parallel
+    {
+        float offsets[3 * block_size];
 
-            // Check boundaries
-            if ( water_coords[iWater].y < size_neighborhood + 1 ) {
-                water_coords[iWater].y = water_coords[iWater].y - 1 * size_neighborhood + vol_nY;
-            }
-            else if ( water_coords[iWater].y > vol_nY - size_neighborhood ) {
-                water_coords[iWater].y = water_coords[iWater].y - vol_nY + 1 * size_neighborhood;
-            }
+#pragma omp for
+        for ( std::size_t iWaterBlock = 0; iWaterBlock < num_blocks; ++iWaterBlock ) {
 
-            // Check boundaries
-            if ( water_coords[iWater].z < size_neighborhood + 1 ) {
-                water_coords[iWater].z = water_coords[iWater].z - 1 * size_neighborhood + vol_nZ;
-            }
-            else if ( water_coords[iWater].z > vol_nZ - size_neighborhood ) {
-                water_coords[iWater].z = water_coords[iWater].z - vol_nZ + 1 * size_neighborhood;
-            }
-        }
-    }
-    else {
-#pragma omp parallel for num_threads(local_threads) private(dr, dx, dy, dz, my_rand)
-        for ( long iWater = 0; iWater < number_of_waters; iWater++ ) {
+            // Local stream for generating this block - ensures that result is independent of thread count
+            // Uses "block-splitting" strategy for parallel generation
+            VSLStreamStatePtr block_stream;
+            vslCopyStream(&block_stream, stream);
+            vslSkipAheadStream(block_stream, iWaterBlock * block_size);
+            vsRngGaussian(VSL_RNG_METHOD_GAUSSIAN_ICDF, block_stream, 3 * block_size, offsets, 0.0, random_sigma);
```

Comment:
Batch up the processing into blocks of size 1024, to avoid using too much temporary memory for holding the generated random numbers. Use MKL to generate random numbers in a vectorized fashion.